Twitter api

sbatson5 opened the pull request:
# Why

This PR adds OAuth registration through Twitter, recording the user's information and updating it when appropriate.

Front-end PR: https://github.com/sbatson5/movie_wager_client/pull/6

Commit:
add user model

add more test coverage

address PR feedback

Commit:
add twitter oauth logic

Commit:
add config secret file

revise tests

revise config imports

Commit:
add session cookie upon login

add better case logic for session

add test for session controller

properly do sessions

michaelpelletier commented on README.md:
```diff
@@ -22,6 +22,19 @@ config :movie_wager_api, MovieWagerApi.Repo,
 
+## Setting up Twitter Authentication
+
+The `dev.secret.exs` and `test.secret.exs` files also include information for the twitter OAuth API.
```
Does this need to be more clear about where that information comes from? I have no idea offhand what a Twitter consumer key is.

sbatson5 replied:
:thinking: possibly. I can add a "ask Scott for these credentials" or link to the twitter dev docs if you want to roll your own.

SViccari replied:
One possibility is to set the dev environment credentials in the `config.exs` file. You already have `import_config "prod.secret.exs"` in the prod.exs file so any extwitter values set in the config file will be overridden by the values set in the `prod.secret.exs` file.

That would allow dev twitter credentials to JustWork™ while still being easily overwritten for production.

SViccari replied:
Since this isn't a private project, it's probably not smart to store the dev credentials in source.  Ignore my comment above :)

RaisedByWolves replied:
We have an secret.example.exs file, correct? We should add in the comments to look to that file if there are questions.
We should also make sure that file is up to date with everything that needs to be specified and enough comments for explanations.

michaelpelletier commented on priv/repo/migrations/20161128200526_create_user.exs:
```diff
@@ -0,0 +1,16 @@
+      add :screen_name, :string, null: false
+      add :profile_image_url, :string
+      add :twitter_id, :integer, null: false
+      add :name, :string
```
What's the difference between name and screen_name and twitter_id?

sbatson5 replied:
Name is the literal name i.e. "Scott Batson."  Screen name is your twitter handle "@ScottJBatson" and twitter_id is a unique identifier twitter uses as you can change your screen name or name at any time.

michaelpelletier commented on web/controllers/twitter_controller.ex:
```diff
@@ -0,0 +1,57 @@
+      {:ok, access_token} ->
+        # TODO: pull consumer info from system
+        ExTwitter.configure(
+          consumer_key: "t5uCiD2SrYK6Zc9AuILSly7C5",
```
Probably you want to alter this before this is merged.

sbatson5 replied:
Yeah, hence the `TODO`

michaelpelletier replied:
Don't hence me, mofo. :P

michaelpelletier commented on web/models/user.ex:
```diff
@@ -0,0 +1,29 @@
+    field :screen_name, :string
+    field :profile_image_url, :string
+    field :twitter_id, :integer
+    field :name, :string
```
Like I said above, I don't think it's super clear what these fields are - I sort of had to read the entire thing to make sense of them.

1. If `twitter_id` is what we use to populate the user id, should that field just be `id`?
2. I'm more in favour of `user_name` and `display_name` over `screen_name` and `name`.

sbatson5 replied:
All of these fields are provided by twitter, so I'd prefer to stick with their naming convention. I'm not sure I want to bind our id to the one provided by twitter.

michaelpelletier replied:
That makes sense if Twitter is the only way that people will ever log in and you don't intend to support a) actual accounts, or b) alternative signon accounts like Facebook. Not sure if you're intending to expand it outward.

I'm also not crazy about our naming conventions suffering because theirs are stupid, but I do get the benefit of consistency.

SViccari replied:
bleh, I was also hesitant about storing first and last name in `name` but I now realize you did it to make it easier to work with Twitter. I have a similar question as @michaelpelletier, is Twitter the only way for a user to register/authenticate? Setting these fields on the user table may make life harder should you wish to add FB or traditional email/password registration.

sbatson5 replied:
my plan right now is to just stick with Twitter. That was the scope of the original game and I'd hate to over optimize this early.

michaelpelletier replied:
That seems reasonable, then. It can always be expanded later.

michaelpelletier reviewed (commented):
Looks great overall! I had a few minor comments.

SViccari commented on config/test.secret.example.exs:
```diff
@@ -4,3 +4,11 @@ use Mix.Config
 #   username: "postgres",
 #   password: "postgres"
+
+# Replace with your twitter app credentials
```
What do you think of removing this addition? I don't believe you'll ever want to make an actual API call in test mode.

sbatson5 replied:
I was actually having test failures when I didn't have this... How would I stub that information?

SViccari replied:
yeah, taking a second pass at this, I realize my comment was incorrect and best ignored :) I think what you have is great and if you ever wanted to remove the "ask Scott for credentials" form the Readme, we could tackle that later on as this project is primarily for two/three people.

SViccari commented on priv/repo/migrations/20161128200526_create_user.exs:
```diff
@@ -0,0 +1,16 @@
+    create table(:users) do
+      add :screen_name, :string, null: false
+      add :profile_image_url, :string
+      add :twitter_id, :integer, null: false
```
Is a user required to register via twitter? Is there existing data that will need to be migrated over to meet the `null: false` constraint?

sbatson5 replied:
There is no existing user data yet as this is the creation step of users table. Right now, I think just Twitter is fine. I think adding other social media links is out of scope of the project.

SViccari commented on test/controllers/session_controller_test.exs:
```diff
@@ -0,0 +1,16 @@
+    test "it returns a 200 when given a user is logged in", %{conn: conn} do
+      conn = sign_in(conn, insert(:user))
+      |> get(session_path(conn, :show, "movie-wager"))
+      |> json_response(200)
```
In addition to the 200 status response, what do you think of asserting that the authenticated user is returned as well?

SViccari commented on test/controllers/user_controller_test.exs:
```diff
@@ -0,0 +1,32 @@
+      end
+    end
+
+    test "it returns a 200 when given a valid id", %{conn: conn} do
```
Testing that we receive the correct status code is nice but I think what we really want to confirm is that this endpoint "returns a user given a valid id" :)

SViccari commented on test/models/user_test.exs:
```diff
@@ -0,0 +1,40 @@
+    assert Repo.one(query).screen_name == user.screen_name
+  end
+
+  test "`by_id_or_screen_name` queries by string id's" do
```
"`by_id_or_screen_name` queries by string id" :)

SViccari commented on test/models/user_test.exs:
```diff
@@ -0,0 +1,40 @@
+    assert Repo.one(query).id == 5
+  end
+
+  test "by_id_or_screen_name` queries by screen_name when passed" do
```
looks like "when passed" could be removed

SViccari commented on web/serializers/user_serializer.ex:
```diff
@@ -0,0 +1,5 @@
+defmodule MovieWagerApi.UserSerializer do
+  use JaSerializer
+
+  attributes [:screen_name, :profile_image_url, :twitter_id, :name]
```
🔤

SViccari reviewed (commented):
Looking good!

Commit:
add LoadAccount Plug test

finish session controller test

sbatson5 commented on web/controllers/movie_round_controller.ex:
```diff
@@ -11,7 +11,7 @@ defmodule MovieWagerApi.MovieRoundController do
     json(conn, serialized_rounds)
   end
 
-  def create(conn, stuff = %{"data" =>  %{"attributes" => movie_round_params}}) do
```
Whoops. That shouldn't have been left in 😛

Commit:
update session endpoints to not require id

properly handle application variables

remove unused variables

update migration to allow bigint

add prod url to config

michaelpelletier reviewed (approved):
Looks great!

RaisedByWolves commented on config/dev.exs:
```diff
@@ -28,6 +28,8 @@ config :movie_wager_api, MovieWagerApi.Endpoint,
 # Do not include metadata nor timestamps in development logs
 config :logger, :console, format: "[$level] $message\n"
 
+config :movie_wager_api, :client, url: "http://localhost:4200"
```
Add a comment here on what this is for.

RaisedByWolves replied:
Also this should not be in dev.exs, this should be in secret. Or another config file that is NOT in source.

sbatson5 replied:
Not sure I follow why this shouldn't be in source. It's not a secret that the app runs at localhost:4200 in development. Why do you feel this should be hidden?

RaisedByWolves replied:
Any server configuration (ports, database addresses, login creds) shouldn't be in source. Credentials for the obvious security concerns, the rest because deploying a server should be data/config driven.
The only reason you would want 'safety fallbacks' in source is because someone deployed a server they haven't fully configured, which I think is better to just let it fail with a useful message.

sbatson5 replied:
But the dev environment isn't a server. I agree that we should remove the client url from `prod.exs` but during development, the app will always be running at `localhost:4200`. The only time it wouldn't would be if the dev purposeful ran it on a different port, in which case they should know to update this.

RaisedByWolves commented on config/dev.secret.example.exs:
```diff
@@ -4,3 +4,11 @@ use Mix.Config
 #   username: "postgres",
 #   password: "postgres"
+
+# Replace with your twitter app credentials
```
ah, you're ahead of me. This is enough info, I don't think it needs more comments.

RaisedByWolves commented on config/prod.exs:
```diff
@@ -21,6 +21,8 @@ config :movie_wager_api, MovieWagerApi.Endpoint,
 # Do not print debug messages in production
 config :logger, level: :info
 
+config :movie_wager_api, :client, url: "http://scarybluerobot.com"
```
This shouldn't be in source.

RaisedByWolves commented:
What are we doing for security? No one can randomly post a delete call to the API address and remove users, correct?

sbatson5 commented:
@RaisedByWolves correct, there is no delete endpoint for users right now. The only thing you can do is delete your own session, which you have to be logged in to do.
